### PR TITLE
Proper Charger/BMS implementation

### PR DESCRIPTION
**Pull request by Apehaenger**
Proper implementation following the quick-fixes in #42

- [x] Separate BMS from PowerService
- [x] Add Charge-Voltage, Re-Charge-Voltage, Pre-Charge-Current and Termination-Current handling

## Summary by CodeRabbit

* **New Features**
  * Added a standalone BMS service for battery telemetry, periodic driver updates, and extra JSON telemetry
  * Added charger controls: adjustable charging voltage and selectable recharge-voltage settings

* **Refactor**
  * Split power service into separate service and driver tick paths; charger config now uses robot defaults when unspecified
  * Charger telemetry surfaced in the UI and robot displays; BMS is now optional/conditional on hardware presence

[Review Change Stack](https://app.coderabbit.ai/change-stack/xtech/fw-openmower-v2/pull/47?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/drivers/bms/sabo_bms_driver.cpp [hunk 1]
```diff
@@ -297,10 +280,8 @@ const char* SaboBmsDriver::GetExtraDataJson() const {
                       ",\"serial_number\":%u,\"design_capacity_ah\":%.3f,\"design_voltage_v\":%.3f",
                       (unsigned)data_.serial_number, (double)data_.design_capacity_ah, (double)data_.design_voltage_v);
 
-  chars += chsnprintf(json_buf + chars, sizeof(json_buf) - chars,
-                      ",\"full_charge_capacity_ah\":%.3f,\"remaining_capacity_ah\":%.3f,\"cycle_count\":%u",
-                      (double)data_.full_charge_capacity_ah, (double)data_.remaining_capacity_ah,
-                      (unsigned)data_.cycle_count);
+  // Status
+  chars += chsnprintf(json_buf + chars, sizeof(json_buf) - chars, ",\"status\":\"%u\"", battery_status_);
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/drivers/bms/sabo_bms_driver.cpp:285]: Fixed in next commit

File: src/drivers/charger/bq_2576/bq_2576.cpp [hunk 2]
```diff
@@ -162,11 +162,24 @@ bool BQ2576::readBatteryVoltage(float& result) {
   return true;
 }
 
+bool BQ2576::setChargingVoltage(float voltage_v) {
+  if (vfb_ratio_ <= 0.0f) return false;
+
+  float vfb_mv = voltage_v * 1000.0f * vfb_ratio_;
+  vfb_mv = std::max(1504.0f, std::min(1566.0f, vfb_mv));  // Clamp to 1504mV - 1566mV
+  uint16_t reg_val = static_cast<uint16_t>((vfb_mv - 1504.0f) / 2.0f);
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/drivers/charger/bq_2576/bq_2576.cpp:184]: Not necessary. If one requests infinite voltage he only get 1566.0f.
What's the problem with that.

File: src/drivers/charger/bq_2576/bq_2576.hpp [hunk 3]
```diff
@@ -14,7 +14,15 @@ using CHARGER_STATUS = ChargerDriver::CHARGER_STATUS;
...
-  explicit BQ2576(float r_ac_sense = 0.0f) : ChargerDriver(), r_ac_sense_(r_ac_sense) {
+  // Internal FBG pull-down resistor to PGND, typic 33 Ohm (max. 55 Ohm)
+  static constexpr uint32_t INTERNAL_RFBG = 33;
+
+  // r_top: external resistor VBAT->FB (Ohm), r_bot: external resistor FB->FBG (Ohm)
+  // vfb_ratio_ = VFB/VBATREG = (r_bot - RFBG) / (r_bot - RFBG + r_top)
+  explicit BQ2576(uint32_t r_top, uint32_t r_bot, float r_ac_sense = 0.0f)
+      : ChargerDriver(),
+        vfb_ratio_(static_cast<float>(r_bot - INTERNAL_RFBG) / static_cast<float>(r_bot - INTERNAL_RFBG + r_top)),
+        r_ac_sense_(r_ac_sense) {
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/drivers/charger/bq_2576/bq_2576.hpp:25]: It's a robot specific settings only entered by the developers and not by users.
Looks nitpicking to me.... and if becoming nitpicking, wouldn't it be also required to test for negative r_top/r_bot values as well as division by zero catchup?

File: src/drivers/ui/lvgl/sabo/sabo_screen_main.hpp [hunk 4]
```diff
@@ -506,15 +522,24 @@ class SaboScreenMain : public ScreenBase<ScreenId, ButtonId> {
...
+    CHARGER_STATUS charger_status = CHARGER_STATUS::UNKNOWN;
+
+    if (charger_->readAdapterVoltage(adapter_volts)) {
+      adapter_volts = std::round(adapter_volts * 10.0f) / 10.0f;
+    }
+    charger_->readChargeCurrent(charge_current);
+    charger_status = charger_->getChargerStatus();
+
+    // Update docked state detection (using adapter voltage)
     is_docked_ = adapter_volts > 10.0f;
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/drivers/ui/lvgl/sabo/sabo_screen_main.hpp]: Good hint!
See my next commits.

File: src/services/bms_service/bms_service.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include "bms_service.hpp"
+
+#include <cstring>
+
+void BmsService::SetDriver(BmsDriver* bms_driver) {
+  bms_ = bms_driver;
+}
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/services/bms_service/bms_service.cpp:14]: Doesn't get handled ATM in the driver. The driver only set a present_ once configured and detected.
BMS driver change is not possible ATM and also no SetNotPresent().

File: src/services/power_service/power_service.cpp [hunk 6]
```diff
@@ -109,14 +91,33 @@ void PowerService::charger_tick() {
...
+        success &= charger_->setChargingVoltage(robot->Power_GetDefaultChargeVoltage());
+      }
+      if (TerminationCurrent.valid && TerminationCurrent.value > 0) {
+        success &= charger_->setTerminationCurrent(TerminationCurrent.value);
+      } else {
+        success &= charger_->setTerminationCurrent(robot->Power_GetDefaultTerminationCurrent());
+      }
+      if (ReChargeVoltage.valid) {
+        success &= charger_->setReChargeVoltage(static_cast<ChargerDriver::ReChargeVoltage>(ReChargeVoltage.value));
+      } else {
```
[the remark this answers is not in the document]
Comment (Apehaenger) [reply src/services/power_service/power_service.cpp:116]: How cares
